storage/engine: add pebble MVCC{Scan,ReverseScan} benchmarks

petermattis opened the pull request:
Release note: None

petermattis committed:
storage/engine: add pebble MVCC{Scan,ReverseScan} benchmarks

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41802)
<!-- Reviewable:end -->

petermattis committed:
storage/engine: tweak the return value for Iterator.MVCCScan

Tweak the return value for `Iterator.MVCCScan` to return a slice of KV
bufs instead of a single buffer. This allows `pebbleMVCCScanner` to scan
into a slice of bufs, rather than a single buf, which avoids copying
data when we need more space. Effectively, this eliminates a copy of
~1/2 of the results. This optimization isn't possible with RocksDB
because we have to copy the C++ memory that we scanned into to Go
controlled memory.

Old is RocksDB and new is Pebble:

name                                             old speed      new speed      delta
MVCCScan/rows=100/versions=1/valueSize=8-16      30.8MB/s ± 4%  38.3MB/s ± 2%  +24.36%  (p=0.000 n=10+10)
MVCCScan/rows=100/versions=1/valueSize=64-16      205MB/s ± 1%   227MB/s ± 3%  +10.53%  (p=0.000 n=10+10)
MVCCScan/rows=100/versions=1/valueSize=512-16    1.23GB/s ± 1%  1.65GB/s ± 2%  +34.12%  (p=0.000 n=9+10)
MVCCScan/rows=1000/versions=1/valueSize=8-16     42.6MB/s ± 2%  47.9MB/s ± 4%  +12.50%  (p=0.000 n=9+10)
MVCCScan/rows=1000/versions=1/valueSize=64-16     285MB/s ± 1%   319MB/s ± 5%  +11.76%  (p=0.000 n=10+10)
MVCCScan/rows=1000/versions=1/valueSize=512-16   1.59GB/s ± 0%  2.16GB/s ± 3%  +35.44%  (p=0.000 n=9+10)
MVCCScan/rows=10000/versions=1/valueSize=8-16    46.5MB/s ± 0%  54.0MB/s ± 2%  +16.23%  (p=0.000 n=8+9)
MVCCScan/rows=10000/versions=1/valueSize=64-16    316MB/s ± 1%   349MB/s ± 5%  +10.57%  (p=0.000 n=10+10)
MVCCScan/rows=10000/versions=1/valueSize=512-16  1.30GB/s ± 1%  1.99GB/s ± 3%  +52.96%  (p=0.000 n=9+10)

Fixes #41597

Release note: None

petermattis commented:
@itsbilal I added a second commit here which addresses the performance optimization described in #41597. This was actually quite easy to do.

itsbilal reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 9 of 9 files at r3, 8 of 8 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41802#-:-LrtcRkz3NAUh89kn3ki:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajkr, @petermattis, and @sumeerbhola)

___
*[pkg/storage/engine/bench_pebble_test.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41802#-Lrox2Wk4DV6XNU5CP33:-Lrox2Wl4bZ6clKTEdw4:b-l598bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/08f74b6f10b9a811b1c711a19bd8bdc7acb7db33/pkg/storage/engine/bench_pebble_test.go#L64)):*
> ```Go
> 			for _, numVersions := range []int{1, 2, 10, 100} {
> 				b.Run(fmt.Sprintf("versions=%d", numVersions), func(b *testing.B) {
> 					for _, valueSize := range []int{8, 64, 512} {
> ```

BTW, were you able to get any of the versions with 512 bytes of valueSize working with either RocksDB or Pebble? When I tried I kept running into `too many open files` errors.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 51 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41802#-LrtbG7c3dC51rrsl-i8:-LrtbG7c3dC51rrsl-i9:b-mqndjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/8953b9baa9d1ae51441f52f4480e3a6d12b15374/pkg/storage/engine/pebble_mvcc_scanner.go#L51)):*
> ```Go
> 	// large the result set might be.
> 	//
> 	// TODO(itsbilal): Have pebbleResults maintain a slice of exponentially
> ```

This TODO can be removed now?
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 59 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41802#-LrtbLQy9fg-KucY206w:-LrtbLQy9fg-KucY206x:b-42aew8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8953b9baa9d1ae51441f52f4480e3a6d12b15374/pkg/storage/engine/pebble_mvcc_scanner.go#L59)):*
> ```Go
> 		newSize := 2 * cap(p.repr)
> 		if newSize == 0 {
> 			newSize = 16
> ```

Nit: have this as a const as well.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 61 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41802#-LrtbDp52CHWIMt5RTao:-LrtbDp52CHWIMt5RTap:b-fu7228) ([raw file](https://github.com/cockroachdb/cockroach/blob/8953b9baa9d1ae51441f52f4480e3a6d12b15374/pkg/storage/engine/pebble_mvcc_scanner.go#L61)):*
> ```Go
> 			newSize = 16
> 		}
> 		const maxSize = 128 << 20 // 128 MB
> ```

Nit: Move this to the start of the function, right next to kvLenSize


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
TFTR! I'll merge on green.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41802#-:-Lrtg-tK-6PRYg72SqAt:b-pzbia2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajkr, @itsbilal, and @sumeerbhola)

___
*[pkg/storage/engine/bench_pebble_test.go, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/41802#-Lrox2Wk4DV6XNU5CP33:-Lrtcmwf626-q5t_VAMY:bnn3hk2) ([raw file](https://github.com/cockroachdb/cockroach/blob/08f74b6f10b9a811b1c711a19bd8bdc7acb7db33/pkg/storage/engine/bench_pebble_test.go#L64)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

BTW, were you able to get any of the versions with 512 bytes of valueSize working with either RocksDB or Pebble? When I tried I kept running into `too many open files` errors.
</blockquote></details>

Yes, they worked for me. Perhaps you have a different `ulimit -n` setting on your system. Mine says `16384`.
___
*[pkg/storage/engine/pebble_mvcc_scanner.go, line 51 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/41802#-LrtbG7c3dC51rrsl-i8:-LrtfwHb94gNZwaMabzI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8953b9baa9d1ae51441f52f4480e3a6d12b15374/pkg/storage/engine/pebble_mvcc_scanner.go#L51)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

This TODO can be removed now?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
